Fix SubViewportContainer processing Events before other Control-Nodes

Sauermann opened the pull request:
Currently, `SubViewportContainer` propagates all its events in `_input`
1 before all other `Control`-Nodes (fix #39666)
2 and where `rect_position` and `rect_size` of a `Control`-Node do not matter (fix #28833)

fix one of the reasons for #58902
alternative to #55300

MRP for all related Bugs: [SubViewportEventBugs.zip](https://github.com/godotengine/godot/files/10106285/SubViewportEventBugs.zip)

This patch changes `SubViewportContainer`, so that it propagates positional events (events that have a position property) no longer in `_input`, but in `_gui_input` to its `SubVieworts` via `push_input`.

### Event coordinates

Since `_input` and `_gui_input` expect different event-position-coordinate-systems, they now have to be adjusted only with the shrink factor in `_gui_input`.

## Changes to Events

Here is a description of how this patch affects events.

1. Events, that contain a `position`

The `position` of the event (like a mouse-click) specifies the `Control`-Node, they are intended for. Just like in master, these events get delivered to the `Control`-Node, that is located at the `position`. They are no longer delivered automatically to `SubViewports` of other `SubViewportContainer`, that are away from the position. The following graphic shows how event propagation of positional events (green line) changes.

![image](https://user-images.githubusercontent.com/6299227/156404182-91bc2cd0-e305-4a1c-940b-ab2e3b2b6d5b.png)


2. Events, that don't contain a `position`

Those events get sent to the same nodes in the same order as in the current master.

## Additional thoughts

- This bugfix solves problems in a different part of Event handling in comparison to #57894

Update 2022-09-29: Evaluate positional Events explicitely
Update 2022-10-02: Fix coordinate transform in _gui_input

zhehangd commented:
My main concern is that this may create a very confusing event order.
Say you have three ColorRect: R,G,B where G is in a subviewport, like below:
![Screenshot_20220223_183203](https://user-images.githubusercontent.com/9572364/155446243-e8d7cda9-8ad0-4588-94c1-72b1c2ef8dd1.jpg)

With this approach the theoretical order becomes
![Screenshot_20220223_183950](https://user-images.githubusercontent.com/9572364/155447137-164f3f34-14ae-4697-a367-99f1dbf51293.jpg)

I think a natural order should be
![Screenshot_20220223_183956](https://user-images.githubusercontent.com/9572364/155447200-9ec48bc3-341e-4956-a4ec-cc6526b443a3.jpg)

as if the viewport does not exist at all. Firing an `input` event from a `gui_input` callback sounds really weird.

Sauermann commented:
Thanks for your review. The following perspective may give a different view on the perceived weirdness:
A Viewport is a different world and as its own entity, it should know, how to process an incoming event (via push_input) all on its own, without the outer world telling it, in which order to process events. This makes push_input the single point of contact between the two worlds and as such simplifies the Viewport interface. So it is not about "Firing an input event from a gui_input callback", but about pushing the event to the subviewport, so that it handles it on its own.

In your suggested order, the combination "R(gui) -> ... -> G(unhandled)" is the reason for issue #17326, because right after "SubViewportContainer(gui)" the event gets set to handled. This makes the sequence "G(gui) -> G(unhandled) -> ... -> R(gui)" necessary.

Sauermann committed:
Adjust SubViewportContainer event handling

Change SubViewportContainer, so that it processes events with a position property no longer in 'input', but in 'gui_input'.
This fixes the issue, that Nodes within its SubViewport receive MouseButton events before other Control-Nodes.

Sauermann commented:
Together with #57894, this PR fixes several outstanding problems about `SubViewport`-`InputEvent`-propagation. Since these problems are reported in different Bug reports with multiple MRPs, I created a single MRP, that demonstrates all of the problems: 

[SubViewportEventBugs.zip](https://github.com/godotengine/godot/files/9922955/SubViewportEventBugs.zip)

![image](https://user-images.githubusercontent.com/6299227/199564020-c268d483-7dde-4643-a0de-f832553cfbb2.png)

Sauermann commented:
With the goal of a better understanding of the changes from my PRs I have created a documentation graphic for event processing including changes from #57894.

![InputEventsFlowchart drawio](https://user-images.githubusercontent.com/6299227/213918935-c80aa170-5da0-4e4e-8264-9bb53daf01d6.png)

For comparison, here is a graphic of the previous event processing order:

![InputEventsFlowchartOld drawio](https://user-images.githubusercontent.com/6299227/213918936-474004ec-162d-45e1-8e10-73ae0b637cad.png)

reduz commented:
This took a while to understand properly because there is just so much to consider, but at this point I agree it's the best possible way to do it.

akien-mga commented:
Thanks!

stryker313 commented:
Was there ever a variant of this for 3.x? #101885

Sauermann commented:
@stryker313 no, there wasn't. There have also been a few other changes affecting input event propagation, that might affect this PR, so back porting to 3.x might be not trivial.
